### PR TITLE
MTLModel+Difference: initial commit.

### PR DESCRIPTION
**Pull request by arangato**
Computes difference between MTLModel objects.

### CODE REVIEW

File: Mantle/MTLModel+Difference.m [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+@implementation MTLModel (Difference)
+
+- (NSArray<NSString *> *)differenceFrom:(id)object {
+  MTLModel *actual = (MTLModel *)object;
+
+  if(![object isKindOfClass:self.class]) {
+		NSString *reason = [NSString stringWithFormat:@"Expected object of type %@, got : %@",
+																									self.class, [object class]];
+		[[[NSException alloc] initWithName:NSInvalidArgumentException	reason:reason
```
Comment (barakwei) [inline Mantle/MTLModel+Difference.m]: is this is an assert, I think we can use NSAssert, no? seems like this will happen only in bugs.

Comment (arangato) [reply Mantle/MTLModel+Difference.m]: replaced with assert and added test for mismatched type.

**Comment by arangato** [comment]
Fixed and squashed.

File: Mantle.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -16,6 +16,12 @@
 		88080C18160A706900CCABF2 /* NSArray+MTLManipulationAdditions.h in Headers */ = {isa = PBXBuildFile; fileRef = 88080C16160A706900CCABF2 /* NSArray+MTLManipulationAdditions.h */; settings = {ATTRIBUTES = (Public, ); }; };
 		88080C19160A706900CCABF2 /* NSArray+MTLManipulationAdditions.m in Sources */ = {isa = PBXBuildFile; fileRef = 88080C17160A706900CCABF2 /* NSArray+MTLManipulationAdditions.m */; };
 		88080C1D160A719D00CCABF2 /* MTLArrayManipulationSpec.m in Sources */ = {isa = PBXBuildFile; fileRef = 88080C1C160A719D00CCABF2 /* MTLArrayManipulationSpec.m */; };
+		94FDA2D425931E9E0031C911 /* MTLModel+Difference.m in Sources */ = {isa = PBXBuildFile; fileRef = 94FDA2D225931E9E0031C911 /* MTLModel+Difference.m */; };
+		94FDA2D525931E9E0031C911 /* MTLModel+Difference.m in Sources */ = {isa = PBXBuildFile; fileRef = 94FDA2D225931E9E0031C911 /* MTLModel+Difference.m */; };
```
Comment (mkimi-lightricks) [inline Mantle.xcodeproj/project.pbxproj]: what are the different places this file is added to?
(same for all the files below)

File: Mantle/MTLModel+Difference.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+// Created by Maxim Grabarnik.
+
+#import <Mantle/Mantle.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MTLModel (Difference)
+
+/// Returns an array of strings describing the difference between the receiver and the given object.
+/// When the receiver and the given object are equal empty array is returned.
```
Comment (mkimi-lightricks) [inline Mantle/MTLModel+Difference.h]: -> `... **an** empty array is returned`.

File: Mantle/MTLModel+Difference.m [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation MTLModel (Difference)
+
+- (NSArray<NSString *> *)differenceFrom:(id)object {
+  MTLModel *actual = (MTLModel *)object;
+
+	NSAssert2([object isKindOfClass:self.class], @"Expected object of type %@, got : %@",
+					  self.class, [object class]);
```
Comment (mkimi-lightricks) [inline Mantle/MTLModel+Difference.m]: indent.

File: Mantle/MTLModel+Difference.m [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  NSMutableArray<NSString *> *differences = [NSMutableArray array];
+  for (NSString *key in self.class.propertyKeys) {
+    id selfValue = [self valueForKey:key];
+    id modelValue = [actual valueForKey:key];
+
+		BOOL valuesEqual = ((selfValue == nil && modelValue == nil) || [selfValue isEqual:modelValue]);
```
Comment (mkimi-lightricks) [inline Mantle/MTLModel+Difference.m]: indent

File: Mantle/MTLModel+Difference.m [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+- (NSArray<NSString *> *)differenceFrom:(id)object {
+  MTLModel *actual = (MTLModel *)object;
+
+	NSAssert2([object isKindOfClass:self.class], @"Expected object of type %@, got : %@",
+					  self.class, [object class]);
+
+  NSMutableArray<NSString *> *differences = [NSMutableArray array];
+  for (NSString *key in self.class.propertyKeys) {
+    id selfValue = [self valueForKey:key];
+    id modelValue = [actual valueForKey:key];
```
Comment (mkimi-lightricks) [inline Mantle/MTLModel+Difference.m]: expectedValue & actualValue?

File: Mantle/MTLModel+Difference.m [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+		BOOL valuesEqual = ((selfValue == nil && modelValue == nil) || [selfValue isEqual:modelValue]);
+    if (!valuesEqual) {
+      [differences addObject:[NSString stringWithFormat:@"%@:", key]];
+      if ([selfValue respondsToSelector:@selector(differenceFrom:)]) {
+        NSArray<NSString *> * diffLines = [selfValue differenceFrom:modelValue];
```
Comment (mkimi-lightricks) [inline Mantle/MTLModel+Difference.m]: -> `*diffLines`

File: MantleTests/MTLModelDifferenceSpec.m [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	expect(difference).to(equal(@[
+    @"count:",
+    @"|\tExpected: 4",
+    @"|\tActual\t: 5",
+		@"name:",
+		@"|\tExpected: foo",
+		@"|\tActual\t: (null)",
+		@"nestedName:",
+		@"|\tExpected: (null)",
+		@"|\tActual\t: moo"
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: indent

File: MantleTests/MTLModelDifferenceSpec.m [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	]));
+});
+
+it(@"should compute different fields of sub-object", ^{
+	NSDictionary *values1 = @{
+		@"name": @"foo",
+	};
+	NSDictionary *values2 = @{
+		@"name": @"moo",
+	};
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: it doesn't really test the computation of diff in sub-object.

File: MantleTests/MTLModelDifferenceSpec.m [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+	NSError *error = nil;
+	MTLTestModel *submodel1 = [[MTLTestModel alloc] initWithDictionary:values1 error:&error];
+	MTLTestModel *submodel2 = [[MTLTestModel alloc] initWithDictionary:values2 error:&error];
+
+	MTLTestModel *model1 = [[MTLTestModel alloc] initWithDictionary:@{ @"weakModel": submodel1 }
+																														error:&error];
+
+	MTLTestModel *model2 = [[MTLTestModel alloc] initWithDictionary:@{ @"weakModel": submodel2 }
+																														error:&error];
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: indent.

File: MantleTests/MTLModelDifferenceSpec.m [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+	NSArray *difference = [model1 differenceFrom:model2];
+	expect(difference).to(equal(@[
+		@"weakModel:",
+		@"|\tname:",
+		@"|\t|\tExpected: foo",
+		@"|\t|\tActual\t: moo"
+	]));
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: indent should be with 2 spaces, not 4 same as [here](https://github.com/Lightricks/Mantle/blob/develop/MantleTests/MTLJSONAdapterSpec.m#L25-L28)

File: MantleTests/MTLModelDifferenceSpec.m [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+#import "MTLModel+Difference.h"
+#import "MTLTestModel.h"
+
+QuickSpecBegin(MTLModelDifferenceSpec)
+
+it(@"should raise on mismatching argument type", ^{
+	MTLTestModel *model = [[MTLTestModel alloc] init];
+	expect(model).notTo(beNil());
+	expect([model differenceFrom:@1]).to(raiseException());
+
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: remove

File: MantleTests/MTLModelDifferenceSpec.m [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+#import <Quick/Quick.h>
+
+#import "MTLModel+Difference.h"
+#import "MTLTestModel.h"
+
+QuickSpecBegin(MTLModelDifferenceSpec)
+
+it(@"should raise on mismatching argument type", ^{
+	MTLTestModel *model = [[MTLTestModel alloc] init];
+	expect(model).notTo(beNil());
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: Optional:
Personally I don't like this expect, cause it doesn't test anything. IMO it should be an assertion.
Same in all test below.

**Review by mkimi-lightricks — CHANGES_REQUESTED** [review]
FAS

File: MantleTests/MTLModelDifferenceSpec.m (hunk 8 above)
Comment (arangato) [reply MantleTests/MTLModelDifferenceSpec.m]: somehow I got TABS all over these files.

File: Mantle.xcodeproj/project.pbxproj (hunk 2 above)
Comment (arangato) [reply Mantle.xcodeproj/project.pbxproj]: restricted to Mantle-iOS only

File: MantleTests/MTLModelDifferenceSpec.m (hunk 9 above)
Comment (arangato) [reply MantleTests/MTLModelDifferenceSpec.m]: It does. values1 and values2 are used to `MTLModel` objects. These MTLModels are used as values for `MTLTestModel`'s `weakModel` fields. So the test compares 2 `MTLTestModel` objects with different `MTLModel` values of some field.
Maybe the name of the test suggested something else ?

**Comment by arangato** [comment]
fixed and squashed

File: MantleTests/MTLModelDifferenceSpec.m (hunk 9 above)
Comment (mkimi-lightricks) [reply MantleTests/MTLModelDifferenceSpec.m]: my bad, you're 💯 .

File: MantleTests/MTLModelDifferenceSpec.m [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  ]));
+});
+
+it(@"should compute different fields of sub-object", ^{
+  NSDictionary *values1 = @{
+    @"name": @"foo",
+  };
+  NSDictionary *values2 = @{
+    @"name": @"moo",
+  };
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: optional: you can inline these array in place where they're used.

File: MantleTests/MTLModelDifferenceSpec.m [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+  NSError *error = nil;
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m:50]: This `error` is never read from. I suggest to add an assertion that it is `nil` after **all** the initializations.

**Review by mkimi-lightricks — APPROVED** [review]
FAS

File: MantleTests/MTLModelDifferenceSpec.m (hunk 15 above)
Comment (mkimi-lightricks) [reply MantleTests/MTLModelDifferenceSpec.m:50]: same for other tests.

**Comment by arangato** [comment]
FASed

File: MantleTests/MTLModelDifferenceSpec.m [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  ]));
+});
+
+it(@"should compute different fields of sub-object", ^{
+  NSError *error = nil;
+  MTLTestModel *submodel1 = [[MTLTestModel alloc] initWithDictionary:@{ @"name": @"foo" }
+																															 error:&error];
+  NSAssert(!error, @"error must be nil");
+  MTLTestModel *submodel2 = [[MTLTestModel alloc] initWithDictionary:@{ @"name": @"moo" }
+																															 error:&error];
```
Comment (mkimi-lightricks) [inline MantleTests/MTLModelDifferenceSpec.m]: indent

**Review by mkimi-lightricks — COMMENTED** [review]
@arangato 

FAS

File: MantleTests/MTLModelDifferenceSpec.m (hunk 16 above)
Comment (arangato) [reply MantleTests/MTLModelDifferenceSpec.m]: sorry. tabs again.

**Comment by arangato** [comment]
Fixed and squashed

File: MantleTests/MTLModelDifferenceSpec.m (hunk 16 above)
Comment (mkimi-lightricks) [reply MantleTests/MTLModelDifferenceSpec.m]: I suggest you to double check your Xcode configuration (use spaces instead of tabs). In you need help LMK.

Comment (arangato) [reply MantleTests/MTLModelDifferenceSpec.m]: Mantle project uses Tabs (project setting), so any modification got them automatically

**Pull request merged**